zerver/lib/upload.py

```diff
@@ -420,14 +420,14 @@ def construct_public_upload_url_base(self) -> str:
 
         base_path = os.path.dirname(parsed.path)
 
-        url_base = urllib.parse.urlunparse(parsed._replace(path=base_path))
-        return url_base
+        url_base_parsed = parsed._replace(path=base_path)
+        return url_base_parsed
 
     def get_public_upload_url(
         self,
         key: str,
     ) -> str:
-        parsed_base = urllib.parse.urlparse(self.public_upload_url_base)
+        parsed_base = self.public_upload_url_base_parsed
         return urllib.parse.urlunparse(
             parsed_base._replace(path=os.path.join(parsed_base.path, key))
```

Comment:
While I think this works, it feels fragile.  Could we just use `urljoin` with the base and key?

Reply:
@timabbott We could use `urljoin`, but isn't it even more fragile due to its kinda subtle semantics with respect to `/` characters?

```
In [9]: urljoin("https://example.com/test/", 'xyz')
Out[9]: 'https://example.com/test/xyz'

In [10]: urljoin("https://example.com/test", 'xyz')
Out[10]: 'https://example.com/xyz'

In [11]: urljoin("https://example.com/test/", '/xyz')
Out[11]: 'https://example.com/xyz'
```

Reply:
We could also use `furl` which is more reasonable, but that would be a new dependency.

```
>>> furl.furl("https://example.com/test/", 'xyz').url
'https://example.com/test/?xyz'
>>> furl.furl("https://example.com/test/").add(path='xyz').url
'https://example.com/test/xyz'
>>> furl.furl("https://example.com/test").add(path='xyz').url
'https://example.com/test/xyz'
>>> furl.furl("https://example.com/test").add(path='/xyz').url
'https://example.com/test/xyz'
>>> furl.furl("https://example.com/test/").add(path='/xyz').url
'https://example.com/test//xyz'
```

Reply:
There’s nothing fragile about `urljoin`. It works exactly like browsers do:
* On `https://example.com/test/`, `<a href="xyz">` links to `https://example.com/test/xyz`,
* On `https://example.com/test`, `<a href="xyz">` links to `https://example.com/xyz`,
* On `https://example.com/test/`, `<a href="/xyz">` links to `https://example.com/xyz`.

Reply:
Yes it's consistent with browser behavior and I assume the relevant RFCs, but my argument is that this doesn't align with the intuition of what we want to do here - the operation we want to do is "append this to the path in the base URL" - and thus it's very counterintuitive that the outcome of the operation is completely not that if slashes are positioned a certain way. By fragile I meant "very easy to  get wrong". `furl`'s `.add()` better reflects what we want to do, but yeah I don't think we want to add the dependency just for this, so it'd be the way to go only if we decided we want to prefer using it in the codebase in general